### PR TITLE
New zero slowdown linear advance algorithm 

### PR DESCRIPTION
**Pull request by dbuezas**
### Description

Linear advance reduces the max acceleration so that the la_advance_rate is kept below e_jerk. The existing implementation also suffers from some degree of quality loss due to it not accounting for the discontinuities in pressure due to both movement jerk and changes in line width (i.e the ratio of e_steps to step_count).

This WIP PR removes that restriction by handling linear advance acceleration in its own isr, and allowing it to not be perfectly in synch with movement. This should theoretically reduce quality, but in practice it doesn't seem to be a problem. In fact, since this algorithm does account for movement jerk and line width changes, the resulting quality is improved beyond the original existing algorithm, and that at higher printing speeds.

The algorithm is as follows:
* Compute the target pressure (in steps) as `current_rate * k`
* Compare with current pressure and accelerate/decelerate extruder to match it.
* When accelerating the extruder, avoid overshooting by ensuring that the extruder will be able to break in time to match the target with zero speed.
* When loading a new block rescale the current (residual) nozzle pressure to the new movement step rate to extrusion step rate ratio. (This handles the aforementioned discontinuities)

This is work in progress.
Optimisation so it can run in slower machines is pending (and ideas very welcome)

### Requirements

Definitely a 32 bit board. For now, maybe even one with a floating point unit

### Benefits

Better quality, faster prints

### Configurations

`#define LA_ZERO_SLOWDOWN`

### Related Issues

https://github.com/MarlinFirmware/Marlin/pull/27259

### CODE REVIEW

**Comment by dbuezas** [comment]
image
**TL;DR: new algorithm is faster and has higher quality.**


TOP LEFT: 300mm/s 4000mm/s2, K=0, 30 mins 
TOP RIGHT: ~82mm/s K=0.15 (OLD ALGORITHM limits to ~1500mm/s2). 48 minutes 
BOTTOM LEFT:  300mm/s 4000/s2, K=0.09, 30 mins.  (yes k changes a lot with high speeds, non linear!)
BOTTOM RIGHT: ~82mm/s 3000mm/s2, k=0.15, 37 mins
There's a printing artefact on the top right corner of the inner hexagon on the old algorithm. It's gone with the new algorithm. I think that's the unaccounted for movement jerk and/or line width changes
Discoloration due to lower temperatures on the bottom left makes it look as if there was more blobbing than there is.

**Comment by ThomasToka** [comment]
Edit: Found it. My error. Will try it now.

**Comment by thinkyhead** [comment]
One potential AVR optimization is to use 16-bit integers for all planner/stepper coordinates on machines that don't require more than 65536 steps to move the full range of X, Y, Z.

**Comment by dbuezas** [comment]
I'm also thinking about how to make a planer for this and how to avoid it lagging slightly by adding look ahead of the next block. I'm working on an algorithm that finds when to accelerate and break to minimize the maximum error between ideal and result, but haven't succeeded yet.

Klipper's moving window averaging is clever but it underutilizes the available extruder max acceleration everywhere except on the more abrupt jumps.

File: Marlin/src/module/stepper.cpp [hunk 1]
```diff
@@ -2843,7 +2870,39 @@ hal_timer_t Stepper::block_phase_isr() {
...
+         * 1. [x] compensate for jerk and line width changes
+         * 2. [x] keep computations in movement step scale
+         * 3. [ ] precompute a_max in movement step scale and inter block correction factors in the planner
+         * 4. [ ] use int arithmetic
+         */
+        if (current_block->step_event_count != 0) {
+          curr_step_rate = current_block->initial_rate;
+          const float old_xy_to_e_steps = xy_to_e_steps;
+          if (current_block->steps.e) {
+            xy_to_e_steps = float(current_block->steps.e) / float(current_block->step_event_count);
```
Comment (XDA-Bam) [inline Marlin/src/module/stepper.cpp:2884]: This var is called `xy_to_e_steps`, but technically computes the inverse. Not critical, but I think you can actually save one division, if you define it as it's named. Go
```c++
          xy_to_e_steps = float(current_block->step_event_count) / float(current_block->steps.e);
```
here and follow up with
```c++
          const float xy_to_e_steps_ratio = xy_to_e_steps / old_xy_to_e_steps;
          current_la_step_count = current_la_step_count * xy_to_e_steps_ratio;
          current_la_step_rate = current_la_step_rate * xy_to_e_steps_ratio;
          a_max = float(planner.max_acceleration_steps_per_s2[E_AXIS + E_INDEX_N(extruder)]) * xy_to_e_steps;
```
at lines 2896 - 2899.

File: Marlin/src/module/stepper.cpp [hunk 2]
```diff
@@ -2856,6 +2915,80 @@ hal_timer_t Stepper::block_phase_isr() {
...
+      }
+      else
+        target_la_step_count = 0; // (de)retration may start with non-zero current_la_step_rate and/or count. This needs to be gradually compensated for
+
+      const float distance_to_target = target_la_step_count - current_la_step_count,
+                  one_shot_v = distance_to_target * dt_inv;
+      float a = ABS(one_shot_v - current_la_step_rate) * dt_inv;
+      NOMORE(a, a_max);
+
+      float stopping_distance = current_la_step_rate * current_la_step_rate / (2 * a_max);
```
Comment (XDA-Bam) [inline Marlin/src/module/stepper.cpp:2962]: Could this be written using `POW(current_la_step_rate, 2)` instead of `current_la_step_rate * current_la_step_rate`?

Comment (dbuezas) [reply Marlin/src/module/stepper.cpp:2962]: Absolutely, is it more readable or also more efficient?

File: Marlin/src/module/stepper.cpp (hunk 1 above)
Comment (dbuezas) [reply Marlin/src/module/stepper.cpp:2884]: My intent was that that's the factor to convert xy steps to e steps:
`xy_to_e = total_e/total_xy`

So then I can multiply xy steps by it to get e steps.

Regardless, I'll take a look at storing the inverse as you say to do less divisions.

File: Marlin/src/module/stepper.cpp (hunk 2 above)
Comment (XDA-Bam) [reply Marlin/src/module/stepper.cpp:2962]: I did not test it. I assume that the compiler will spit out the same assembler code in most cases. But `POW` explicitly relies on the HAL, so - depending on the arch - it _might_ be faster than doing `x * x` in some cases. But mostly, it just feels like the right command to use here 😉

**Comment by dbuezas** [comment]
I switched gears and followed a completely different approach in https://github.com/MarlinFirmware/Marlin/pull/27709
